jetty-util/src/main/java/org/eclipse/jetty/util/resource/ZipFsResource.java

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    @Override
+    public boolean exists()
+    {
+        // If we passed the constructor, this is known to exist
+        return true;
+    }
+
+    @Override
+    public boolean isDirectory()
+    {
+        // a ZipFsResource is always a directory (never a file)
+        return true;
+    }
+
+    @Override
+    public Iterator<String> iterator()
+    {
+        try
+        {
+            DirectoryStream<Path> dirStream = Files.newDirectoryStream(rootPath);
+            return new PathToStringIterator(dirStream);
+        }
+        catch (IOException e)
+        {
+            LOG.warn("Unable to get iterator for: " + this, e);
+            return Collections.emptyIterator();
+        }
+    }
+
+    @Override
+    public long lastModified()
+    {
+        try
+        {
+            return Files.getLastModifiedTime(zipFsPath).toMillis();
+        }
+        catch (IOException ignore)
+        {
+            LOG.ignore(ignore);
+            return -1;
+        }
+    }
+
+    @Override
+    public long length()
+    {
+        try
+        {
+            return Files.size(zipFsPath);
+        }
+        catch (IOException ignore)
+        {
+            LOG.ignore(ignore);
+            return -1;
+        }
+    }
+
+    @Override
+    public URL getURL()
+    {
+        try
+        {
+            return zipFsUri.toURL();
+        }
+        catch (MalformedURLException e)
+        {
+            throw new RuntimeException("Unable to get url for " + this);
+        }
+    }
+
+    @Override
+    public File getFile() throws IOException
+    {
+        return zipFsPath.toFile();
+    }
+
+    @Override
+    public String getName()
+    {
+        return zipFsPath.toString();
+    }
+
+    @Override
+    public InputStream getInputStream() throws IOException
+    {
+        throw new IOException("Cannot open InputStream for " + zipFs.provider().getClass().getName());
+    }
+
+    @Override
+    public ReadableByteChannel getReadableByteChannel() throws IOException
+    {
+        throw new IOException("Cannot open ReadableByteChannel for " + zipFs.provider().getClass().getName());
+    }
+
+    @Override
+    public boolean delete() throws SecurityException
+    {
+        throw new UnsupportedOperationException("Not supported by " + ZipFsResource.class.getSimpleName());
+    }
+
+    @Override
+    public boolean renameTo(Resource resource) throws SecurityException
+    {
+        throw new UnsupportedOperationException("Not supported by " + ZipFsResource.class.getSimpleName());
+    }
+
+    @Override
+    public String[] list()
+    {
+        ArrayList<String> entries = new ArrayList<>();
+        iterator().forEachRemaining((entry) -> entries.add(entry));
+        return entries.toArray(new String[0]);
+    }
+
+    @Override
+    public Resource addPath(String s) throws IOException, MalformedURLException
+    {
+        Path path = rootPath.resolve(s);
+        return new PathResource(path);
```

Comment:
Oh I'm confused now?  How can we get a `PathResource` out of a `ZipFsResource`?

Surely the result of:
```
    new ZipFsResource("jar:file:///some/path/file.zip!/foo").addPath("bar");
```
is the same as 
```
    new ZipFsResource("jar:file:///some/path/file.zip!/foo/bar");
```
Confused?

Reply:
Those are not ZipFs FileSystem references or valid for ZipFsResource, those are both paths.